# Possible outliers are in 

# 1. offer_valid_from , offer_valid_to 
# 2. available_cnt
# 3. sellings_price
# 4. valid_offer_flag
# 5. breakfast_included_flag

In [38]:
# import libraries 

import pandas as pd, numpy as numpy
import matplotlib.pyplot as plt
import seaborn as sns
%pylab inline
from datetime import datetime

Populating the interactive namespace from numpy and matplotlib


In [21]:
df = pd.read_csv('offer.csv')

In [23]:
df.columns = ['id',
			'hotel_id',
			'currency_id',
			'source_system_code', 
			'available_cnt', 
			'sellings_price',
			'checkin_date',
			'checkout_date',
			'valid_offer_flag', 
			'offer_valid_from',
			'offer_valid_to', 
			'breakfast_included_flag', 
			'insert_datetime' ]
df.head(3)

id  hotel_id  currency_id                source_system_code  available_cnt  \
0   6      11.0          1.0  e0309fe08002dddd49518539a92118dc            2.0   
1   8      11.0          1.0  e0309fe08002dddd49518539a92118dc            5.0   
2   9      11.0          1.0  e0309fe08002dddd49518539a92118dc            8.0   

   sellings_price checkin_date checkout_date  valid_offer_flag  \
0            99.0   2012-11-21    2012-11-22                -1   
1           120.0   2012-11-21    2012-11-22                -1   
2           120.0   2012-11-20    2012-11-21                -1   

      offer_valid_from       offer_valid_to  breakfast_included_flag  \
0  1900-01-01 00:00:00  2999-12-31 00:00:00                       -1   
1  1900-01-01 00:00:00  2999-12-31 00:00:00                       -1   
2  1900-01-01 00:00:00  2999-12-31 00:00:00                       -1   

       insert_datetime  
0  2012-11-20 17:09:09  
1  2012-11-20 17:09:58  
2  2012-11-20 17:11:36

In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5290818 entries, 0 to 5290817
Data columns (total 15 columns):
id                         int64
hotel_id                   float64
currency_id                float64
source_system_code         object
available_cnt              float64
sellings_price             float64
checkin_date               object
checkout_date              object
valid_offer_flag           int64
offer_valid_from           object
offer_valid_to             object
breakfast_included_flag    int64
insert_datetime            object
period_valid               int64
valid_period               int64
dtypes: float64(4), int64(5), object(6)
memory usage: 605.5+ MB


# 1. offer_valid_from , offer_valid_to

In [26]:
df[['offer_valid_from','offer_valid_to']].head(3)

offer_valid_from       offer_valid_to
0  1900-01-01 00:00:00  2999-12-31 00:00:00
1  1900-01-01 00:00:00  2999-12-31 00:00:00
2  1900-01-01 00:00:00  2999-12-31 00:00:00

In [89]:
# only get year part from offer_valid_to, and offer_valid_from, then get the difference 
df['valid_period'] = df['offer_valid_to'].apply(lambda x: int((str(x))[:4])) - df['offer_valid_from'].apply(lambda x: int((str(x))[:4]))

In [88]:
df.groupby('valid_period').count()['id']

period_valid
-2             2
-1             1
 0       5290628
 1            78
 985          24
 986          32
 1099         53
Name: id, dtype: int64

### Most valid_period are within 1 year, but some of them are unbelievable long ( 985, 986 1099 year), and minus so I recognize them as outliers here 

# 2. available_cnt

In [95]:
df.groupby('available_cnt').count()['id']

available_cnt
-2.0            1
-1.0           10
 0.0       351993
 1.0      3550603
 2.0       143582
 3.0        96801
 4.0        69179
 5.0       153525
 6.0        29180
 7.0        32191
 8.0        39912
 9.0        50342
 10.0      128249
 11.0        8179
 12.0        8761
 13.0        7808
 14.0       10662
 15.0       14185
 16.0        8263
 17.0        8671
 18.0       12639
 19.0       24135
 20.0       35452
 21.0        4205
 22.0        4739
 23.0        6336
 24.0        6980
 25.0        5321
 26.0        3974
 27.0        3871
           ...   
 288.0         12
 289.0         12
 290.0         15
 291.0         18
 292.0         20
 293.0         21
 294.0         22
 295.0         19
 296.0         26
 297.0         48
 298.0         28
 299.0        107
 300.0         36
 975.0          6
 980.0          3
 981.0          3
 986.0          2
 987.0          2
 988.0          1
 989.0        387
 990.0        344
 991.0        517
 992.0        144
 993.0        

### available_cnt should > 0 since this means Number of rooms available

# 3. sellings_price

In [97]:
df.groupby('sellings_price').count()['id']

sellings_price
-1.900000e+06      1
-6.801984e+04      3
-6.551633e+04      4
-6.532770e+04      3
-6.495373e+04      1
-6.458964e+04      1
-6.174810e+04      4
-6.114200e+04      5
-5.833646e+04      1
-5.816850e+04      1
-5.743898e+04      1
-5.709500e+04      1
-5.460482e+04      3
-5.219678e+04      1
-5.213600e+04     11
-4.936162e+04      3
-4.921950e+04      1
-4.634100e+04      1
-4.433411e+04      1
-4.421290e+04      1
-4.409728e+04      1
-4.397672e+04      1
-4.015060e+04      1
-3.959756e+04      1
-3.951000e+04      2
-3.937560e+04      1
-3.930035e+04      1
-3.731500e+04      2
-3.687930e+04      1
-3.684408e+04      1
                ... 
 4.118718e+08      1
 4.122626e+08      1
 4.142163e+08      7
 4.374440e+08      1
 4.376324e+08      1
 4.395024e+08      1
 4.564879e+08      6
 6.495658e+08      2
 7.035089e+08     10
 7.042129e+08      2
 7.507432e+08      1
 8.695436e+08      2
 8.926127e+08      1
 2.147480e+09      4
 2.147484e+09     62
 3.189656e+09      

###  sellings_price should > 0 since these are prices 

# 4. valid_offer_flag

In [98]:
df.groupby('valid_offer_flag').count()['id']

valid_offer_flag
-1      88412
 0    1828518
 1    3373888
Name: id, dtype: int64

### Guess  valid_offer_flag only has status 1 & 0 (yes or no)

# 5. breakfast_included_flag

In [99]:
df.groupby('breakfast_included_flag').count()['id']

breakfast_included_flag
-1      39434
 0    3922992
 1    1328392
Name: id, dtype: int64

###  Guess breakfast_included_flag only has status 1 & 0 (yes or no)